src/auth/src/credentials/mds.rs

```diff
@@ -250,32 +267,47 @@ impl TokenProvider for MDSAccessTokenProvider {
             .into_iter()
             .fold(request, |r, s| r.query(&[("scopes", s)]));
 
-        let response = request.send().await.map_err(errors::retryable)?;
-        // Process the response
-        if !response.status().is_success() {
-            let status = response.status();
-            let body = response
-                .text()
-                .await
-                .map_err(|e| CredentialsError::new(is_retryable(status), e))?;
-            return Err(CredentialsError::from_str(
-                is_retryable(status),
-                format!("Failed to fetch token. {body}"),
-            ));
+        match request.send().await {
+            Ok(response) => {
+                // Process the response
+                if !response.status().is_success() {
+                    let status = response.status();
+                    let body = response
+                        .text()
+                        .await
+                        .map_err(|e| CredentialsError::new(is_retryable(status), e))?;
+                    return Err(CredentialsError::from_str(
+                        is_retryable(status),
+                        format!("Failed to fetch token. {body}"),
+                    ));
+                }
+                let response = response.json::<MDSTokenResponse>().await.map_err(|e| {
+                    let retryable = !e.is_decode();
+                    CredentialsError::new(retryable, e)
+                })?;
```

Comment:
I think there are too many exit paths here. It is unclear why only one gets the detailed error message.

It may be better to refactor this to a function and then apply the detail error outside. Or maybe document why you return a detail error in some cases but not others.

Or maybe create some helpers , something like `with_detailed_error(retryable,e)` and `with_plain_error(retryable, e)` so it becomes self-documenting (or at least clear that we considered each case).

Reply:
That is a good idea, let me create with_detailed_error(retryable,e) and with_plain_error(retryable, e)

Reply:
Moved the decision logic on creating the appropriate error to a single place instead.